Initial

Max-Pochepets opened the pull request:
DAO3

Max-Pochepets committed:
Initial

Mandybur800 commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        return Optional.of(Storage.manufacturersStorage.get(id));
```
avoid NullPointException

Max-Pochepets committed:
probable NPE fixed

Max-Pochepets committed:
probable NPE fixed(2)

Max-Pochepets committed:
Added Travis

Sofasmile commented on pom.xml:
```diff
@@ -9,8 +9,52 @@
+        </pluginManagement>
+    </build>
+
 </project>
```
add empty line at the end of the file

Sofasmile commented on src/main/java/crud/dao/abstraction/CarDao.java:
```diff
@@ -0,0 +1,7 @@
+import crud.dao.abstraction.src.Dao;
+import crud.model.Car;
+
+public interface CarDao extends Dao<Car> {
```
for this hw it is redundant, let's remove it from PR

Sofasmile commented on src/main/java/crud/dao/abstraction/DriverDao.java:
```diff
@@ -0,0 +1,7 @@
+import crud.dao.abstraction.src.Dao;
+import crud.model.Driver;
+
+public interface DriverDao extends Dao<Driver> {
```
the same about DriverDao

Sofasmile commented on src/main/java/crud/dao/abstraction/src/Dao.java:
```diff
@@ -0,0 +1,16 @@
+
+    Optional<T> get(Long id);
+
+    Map<Long, T> getAll();
```
```suggestion
    List<T> getAll();
```

Sofasmile commented on src/main/java/crud/storage/Storage.java:
```diff
@@ -0,0 +1,34 @@
+    public static void addManufacturer(Manufacturer manufacturer) {
+        manufacturerId++;
+        manufacturer.setId(manufacturerId);
+        manufacturersStorage.put(manufacturer.getId(), manufacturer);
```
```suggestion
        manufacturersStorage.put(manufacturer.getId(), manufacturer);
        return manufacturer;
```

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,43 @@
+        if (Storage.manufacturersStorage.containsValue(manufacturer)) {
+            return manufacturer;
+        }
+        Storage.addManufacturer(manufacturer);
+        return Storage.manufacturersStorage.get(manufacturer.getId());
```
```suggestion
        return Storage.addManufacturer(manufacturer);
```

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,43 @@
+public class ManufacturerDaoImpl implements ManufacturerDao {
+    @Override
+    public Manufacturer create(Manufacturer manufacturer) {
+        if (Storage.manufacturersStorage.containsValue(manufacturer)) {
```
this checking is redundant

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -0,0 +1,43 @@
+        if (Storage.manufacturersStorage.containsKey(manufacturer.getId())) {
+            Storage.manufacturersStorage.put(manufacturer.getId(), manufacturer);
+        }
+        return create(manufacturer);
```
now your method works as `saveOrUpdate()` method. In the future, it will be ok, but now, let's do only update without save in method update()

Sofasmile commented on src/main/java/crud/lib/DaoImpl.java:
```diff
@@ -0,0 +1,8 @@
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface DaoImpl {
```
```suggestion
public @interface Dao {
```

Max-Pochepets replied:
https://github.com/Max-Pochepets/jv-create-annotation/pull/2#discussion_r554951749

Sofasmile commented on src/main/java/crud/lib/ServiceImpl.java:
```diff
@@ -0,0 +1,8 @@
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface ServiceImpl {
```
```suggestion
public @interface Service {
```

Max-Pochepets replied:
https://github.com/Max-Pochepets/jv-create-annotation/pull/2#discussion_r554951749

Sofasmile commented on src/main/java/crud/service/abstraction/CarService.java:
```diff
@@ -0,0 +1,7 @@
+import crud.model.Car;
+import crud.service.abstraction.src.Service;
+
+public interface CarService extends Service<Car> {
```
the same, let's remove CarService and DriverService from this PR

Sofasmile commented on src/main/java/crud/service/abstraction/src/Service.java:
```diff
@@ -0,0 +1,15 @@
+
+    T get(Long id);
+
+    Map<Long, T> getAll();
```
```suggestion
    List<T> getAll();
```

Sofasmile commented on src/main/java/crud/service/impl/ManufacturerServiceImpl.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public Manufacturer get(Long id) {
+        if (manufacturerDao.get(id).isEmpty()) {
+            return null;
+        }
```
this checking is redundant

Sofasmile commented on src/main/java/crud/storage/Storage.java:
```diff
@@ -0,0 +1,34 @@
+public class Storage {
+    public static Map<Long, Car> carsStorage = new HashMap<>();
+    public static Map<Long, Driver> driversStorage = new HashMap<>();
+    public static Map<Long, Manufacturer> manufacturersStorage = new HashMap<>();
```
```suggestion
    public static Map<Long, Manufacturer> manufacturers = new HashMap<>();
```
the same with others

Max-Pochepets replied:
https://github.com/Max-Pochepets/jv-create-annotation/pull/2#discussion_r554951749

Max-Pochepets committed:
Refactor:

deleted redundant files;
changed behaviour of the storage;
storage is now List;
improved DAO and Service logic

Sofasmile commented on src/main/java/crud/dao/abstraction/src/Dao.java:
```diff
@@ -1,16 +1,16 @@
     T create(T element);
 
-    Optional<T> get(Long id);
+    Optional<T> get(long id);
```
why do we change wrapper to primitive?
let's return to the previous version

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -4,40 +4,34 @@
-    public Optional<Manufacturer> get(Long id) {
-        return Optional.ofNullable(Storage.manufacturersStorage.get(id));
+    public Optional<Manufacturer> get(long id) {
+        return Optional.ofNullable(Storage.manufacturersStorage.get((int) id));
```
will this method work correctly?

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -4,40 +4,34 @@
-            Storage.manufacturersStorage.put(manufacturer.getId(), manufacturer);
-        }
-        return create(manufacturer);
+        Storage.manufacturersStorage.set((int) manufacturer.getId(), manufacturer);
```
the same about it? does this method work correctly?

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -4,40 +4,34 @@
-    public boolean delete(Long id) {
-        return Storage.manufacturersStorage.remove(id, Storage.manufacturersStorage.get(id));
+    public boolean delete(long id) {
+        return Storage.manufacturersStorage.remove(Storage.manufacturersStorage.get((int) id));
```
read about method `removeIf`

Sofasmile commented on src/main/java/crud/model/Car.java:
```diff
@@ -5,7 +5,7 @@
 
 public class Car {
-    private Long id;
+    private long id;
```
let's return wrapper here

Max-Pochepets committed:
Refactor:

wrappers returned;
get and update methods improved;

Max-Pochepets committed:
storage name changed

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -15,23 +16,26 @@ public Manufacturer create(Manufacturer manufacturer) {
-        Storage.manufacturersStorage.set((int) manufacturer.getId(), manufacturer);
-        return manufacturer;
+        Manufacturer oldManufacturer = manufacturer;
+        Storage.manufacturers.set(manufacturer.getId().intValue() - 1, manufacturer);
```
чи співпадає ід об'єкта з позицією його в лісті?

Max-Pochepets replied:
no, cause indexes goes from 0, but IDs from 1

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -15,23 +16,26 @@ public Manufacturer create(Manufacturer manufacturer) {
-        Storage.manufacturersStorage.set((int) manufacturer.getId(), manufacturer);
-        return manufacturer;
+        Manufacturer oldManufacturer = manufacturer;
+        Storage.manufacturers.set(manufacturer.getId().intValue() - 1, manufacturer);
```
чи співпадає ід об'єкта з позицією його в лісті?

Sofasmile commented on src/main/java/crud/service/impl/ManufacturerServiceImpl.java:
```diff
@@ -18,8 +18,8 @@ public Manufacturer create(Manufacturer manufacturer) {
-    public Manufacturer get(long id) {
-        return manufacturerDao.get(id).get();
+    public Manufacturer get(Long id) {
+        return manufacturerDao.get(id).orElse(null);
```
```suggestion
        return manufacturerDao.get(id).get();
```

Max-Pochepets committed:
dao update method improved: update now depends strictly on index

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -29,8 +29,8 @@ public Manufacturer create(Manufacturer manufacturer) {
-        Storage.manufacturers.set(manufacturer.getId().intValue() - 1, manufacturer);
+        Manufacturer oldManufacturer = get(manufacturer.getId()).get();
+        Storage.manufacturers.set(Storage.manufacturers.indexOf(oldManufacturer), manufacturer);
         return oldManufacturer;
```
we should return new value in update method

Max-Pochepets committed:
update

Sofasmile commented on src/main/java/crud/dao/impl/ManufacturerDaoImpl.java:
```diff
@@ -29,9 +29,10 @@ public Manufacturer create(Manufacturer manufacturer) {
+        Manufacturer oldManufacturer = get(desiredId).get();
         Storage.manufacturers.set(Storage.manufacturers.indexOf(oldManufacturer), manufacturer);
-        return oldManufacturer;
+        return get(desiredId).get();
```
or `return manufacturer;`
